### PR TITLE
UnnecessaryLet: fix false positive when let is used for destructuring

### LINKED ISSUE #2966: UnnecessaryLet false positive when let is used for destructuring
## Expected Behavior

I don't expect `UnnecessaryLet` to be reported for the following code:

```
val y = x.let { (a, b) -> a + b }
```

That is because, in this case, `let` is necessary for destructuring. If I really, want to get rid of `let` I have to destructure in a `val (a, b) = x` assigment. And to not let `a` and `b` leak in the scope, I'd have to replace `val y = x.let { (a, b) -> a + b}` by `val y = run { val (a, b) = x ; a + b }`. Not really an improvement in my opinion.

## Observed Behavior

An `UnnecessaryLet` is reported for the code above.

## Steps to Reproduce

1. Enable `UnnecessaryLet` rule
2. Write:

```kotlin
fun foo(x: Pair<Int, Int>): Int = x.let { (a, b) -> a + b }
```

3. Run detekt

## Your Environment

* Version of detekt used: 1.11.0-RC2

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #2966

### CODE REVIEW

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLet.kt [hunk 1]
```diff
@@ -110,6 +113,10 @@ private fun KtBlockExpression.hasOnlyOneStatement() = this.children.size == 1
 private fun PsiElement.countVarRefs(varName: String): Int =
     children.sumBy { it.countVarRefs(varName) + if (it.textMatches(varName) && it !is ValueArgument) 1 else 0 }
 
-private fun KtLambdaExpression.countReferences(): Int? {
-    return bodyExpression?.countVarRefs(firstParameter?.text ?: IT_LITERAL)
+private fun KtLambdaExpression.countReferences(): Int {
+    val bodyExpression = bodyExpression ?: return 0
+    val firstParameter = firstParameter
```
Comment (schalkms) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLet.kt:118]: I think this one can be removed safely.

Comment (schalkms) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLet.kt:118]: `val firstParameter = firstParameter`

**Comment by schalkms** [comment]
👏

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLet.kt [hunk 2]
```diff
@@ -110,6 +113,10 @@ private fun KtBlockExpression.hasOnlyOneStatement() = this.children.size == 1
...
+    val destructuringDeclaration = firstParameter?.destructuringDeclaration
+    return destructuringDeclaration?.entries?.sumBy { bodyExpression.countVarRefs(it.nameAsSafeName.asString()) }
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLet.kt:120]: Can you divide this expression in 2? Having multiple occurrencies of the elvis operator on the same line makes the code harder to read.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLetSpec.kt [hunk 3]
```diff
@@ -191,6 +191,96 @@ class UnnecessaryLetSpec : Spek({
...
+                
+                fun test(foo: Foo) {
+                    foo.let { (a, _) -> a + a }
+                }
+            """
+                val findings = subject.compileAndLint(content)
+                assertThat(findings).isEmpty()
+            }
+
+            it("does not report when destructuring declarations are used more than once 3") {
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryLetSpec.kt:220]: Please don't use 2,3,... in test names but try to be a bit more specific.
Given that you have a `context("destructuring declarations")` on top you can probably simplify to 

```suggestion
            it("does not report when destructuring with types") {
```

and so on.

**Pull request merged**